Add support for headless in recent (> ~109) chrome versions

japiirainen opened the pull request:
## problem

I recently ran into some problems when running tests with `cuic` in headless mode. In particular I started getting `Could not detect Chrome tabs` exceptions.

This seems to be caused by chromium doing a breaking change recently (version ~109). When started headless,
it seems it has "zero" tabs open, which causes `cuic` to crash in `chrome/launch`.

## solution

This can be fixed by providing an url and `about:blank` is short and failsafe way to provide one. Chromium also recently released 'new' headless mode (https://developer.chrome.com/articles/new-headless/), which seems to also fix this issue.
I decided ***not*** to make it the new default because `playwright` team found it up to 50% slower in some cases.

## After this PR

- `cuic` works in headless mode with chrome versions > 109
- `cuic` supports 'new' headless (https://developer.chrome.com/articles/new-headless/)
- 'old' headless is still used by default since this way we avoid breaking changes +
  `playwright` developers had noticed ~30-50% performance hit with 'new' headless.
  see: https://github.com/microsoft/playwright/issues/21216
- (chrome/launch) :headless parameter now supports these values:
    - false -> same as before
    - true  -> gets translated to 'old'
    - old   -> runs chrome in --headless=old
    - new   -> runs chrome in --headless=new

I did a not very scientific performance experiment via `time`, which suggest that in `cuic` case there seems to be only minimal performance difference between old and new headless modes.

Using old headless

<img width="781" alt="Screenshot 2023-05-19 at 10 06 52" src="https://github.com/milankinen/cuic/assets/61139818/c1409f2c-e2cb-4ed4-931e-8fbf3ffa7213">

Using new headless

<img width="781" alt="Screenshot 2023-05-19 at 10 22 28" src="https://github.com/milankinen/cuic/assets/61139818/cc6e9bcf-3d9e-41d7-a710-ade34b805cbd">

japiirainen commented:
Hmm, CI seems to not get triggered for some reason.. Maybe because of this https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#enabling-workflows-for-forks-of-private-repositories?

milankinen commented on src/clj/cuic/core.clj:
```diff
@@ -168,7 +168,7 @@
   "Stops the current thread execution for the given
    `n` milliseconds."
-  [ms]
+  [^long ms]
```
👍

milankinen commented on doc/chrome.md:
```diff
@@ -17,6 +17,21 @@ reference of the available startup options.
+(def old-headless-chrome (chrome/launch {:headless "old"}))
+
+;; launch in new headless mode
+(def old-headless-chrome (chrome/launch {:headless "new"}))
```
Can we also add a note about previous boolean usage and perhaps add a deprecation notice?

```clj
;; deprecated: equivalent with {:headless "old"} but may be removed in future
(def old-headless-chrome (chrome/launch {:headless true}))
```

japiirainen replied:
Will do!

milankinen commented on src/clj/cuic/chrome.clj:
```diff
@@ -346,20 +353,23 @@
-                          {:user-data-dir         (str data-dir)
-                           :remote-debugging-port port})
-                   (filter second)
+         options (merge (resolve-headless-opts options)
```
I think `resolve-headless-opts` is a little bit misleading naming here since the function applies defaults if the `:headless` flag is set to `false`. I'd either

1. separate headless handling out of function
```clj
(if-some [headless-mode (:headless options)]
  (resolve-headless-defaults headless-mode)
  defaults)
```

2. or rename the function to e.g. `(resolve-defaults options)`

japiirainen replied:
Good catch, That naming was indeed confusing!

milankinen reviewed (commented):
Wow, thank you so much for this PR! 😊 So far looking good, I left some comments about minor issues if it's not a big deal to change those.

milankinen commented:
Hmm for some reason I don't see the "approve workflow run" button anywhere... 🤔

japiirainen commented:
Thanks for the review! I addressed the issues you pointed out in comments. :-)

Weird that CI is not getting triggered. I opened a PR on my fork [here](https://github.com/japiirainen/cuic/pull/1) and CI seems to be passing fine.

milankinen reviewed (approved):
I ran the tests locally and it passes 👍
